apps/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts

```diff
@@ -0,0 +1,588 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as colors from 'colors';
+import * as os from 'os';
+import * as path from 'path';
+import * as semver from 'semver';
+import { FileSystem, InternalError, MapExtensions, NewlineKind } from '@rushstack/node-core-library';
+
+import { AlreadyReportedError } from '../../utilities/AlreadyReportedError';
+import { BaseInstallManager, IInstallManagerOptions } from '../base/BaseInstallManager';
+import { BaseShrinkwrapFile } from '../../logic/base/BaseShrinkwrapFile';
+import { DependencySpecifier } from '../DependencySpecifier';
+import { PackageJsonEditor, DependencyType } from '../../api/PackageJsonEditor';
+import { PnpmWorkspaceFile } from '../pnpm/PnpmWorkspaceFile';
+import { RushConfiguration } from '../../api/RushConfiguration';
+import { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import { RushConstants } from '../../logic/RushConstants';
+import { Stopwatch } from '../../utilities/Stopwatch';
+import { Utilities } from '../../utilities/Utilities';
+import { InstallHelpers } from './InstallHelpers';
+import { CommonVersionsConfiguration } from '../../api/CommonVersionsConfiguration';
+import { RepoStateFile } from '../../api/RepoStateFile';
+
+/**
+ * This class implements common logic between "rush install" and "rush update".
+ */
+export class WorkspaceInstallManager extends BaseInstallManager {
+  public static getCommonWorkspaceKey(rushConfiguration: RushConfiguration): string {
```

Comment:
A workspace key is the key by which a workspace is referred to in the `pnpm-lock.yaml` file, and specifically for PNPM, this is located under the `importers` field. Practically speaking, it's a relative path from the `pnpm-lock.yaml` to the project root. However, I've removed this method because I realized that it's not getting used anymore (yay)